### PR TITLE
Separate WGs and Projects lifecycles Take 2

### PR DESCRIPTION
**Pull request by lehors**
This is a different take on the separation of WGs and Projects which tries to stick to the reality OpenSSF finds itself in and where we have many initiatives called "projects" that are not software focused.

This approach leverages the existing WG lifecycle and leverages the new lifecycle
inherited from CNCF for software focused projects while introducing a new
yet-to-be-defined lifecycle for non-software focused projects.

Signed-off-by: Arnaud J Le Hors <lehors@us.ibm.com>

IMPORTANT NOTE: This PR is an alternative to PR #95. Only one of them should be merged.

Additional Note: If adopted, this approach should be followed by an update of the OpenSSF charter to match this model.

### CODE REVIEW

File: process/software-project-lifecycle.md [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+Sandbox projects should be early-stage projects that the OpenSSF TAC or parent Working Group believes warrant experimentation.
+
+* New software focused projects that are designed to extend one or more OpenSSF Projects with functionality or interoperability libraries.
+* Independent software focused projects that fit the OpenSSF mission and provide potential for a novel approach to existing functional areas (or are an attempt to meet an unfulfilled need)
+* Software focused projects commissioned or sanctioned by the OpenSSF, including initial code for OpenSSF WG collaborations, and "experimental" projects
+* Any software focused project that realistically intends to join OpenSSF Incubation in future and wishes to lay the foundations for that.
+
+#### Project Graduation Process: Sandbox to Incubating
+
+Incubating software focused projects are required to undergo due diligence as a part of the process to move from Sandbox to Incubation. Due Diligence is driven by a TAC or parent WG sponsor, with two weeks for public comment before a vote is called.
```
Comment (AevaOnline) [inline process/software-project-lifecycle.md:70]: akin to my comment on #95, even sandbox projects require legal review / IP transfer to be accepted. Further diligence should be reserved for Incubating, as here described.

File: process/software-project-lifecycle.md [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+To be accepted to incubating stage, a project must meet the sandbox stage requirements plus:
+
+* Document that it is being used successfully in production by at least three independent end users which, in the TAC’s judgement, are of adequate quality and scope.
+* Have a healthy number of committers. A committer is defined as someone with the commit bit; i.e., someone who can accept contributions to some or all of the project.
+* Demonstrate a substantial ongoing flow of commits and merged contributions.
+* Since these metrics can vary significantly depending on the type, scope and size of a project, the TAC has final judgement over the level of activity that is adequate to meet these criteria
+* A clear versioning scheme.
+* Specifications must have at least one public reference implementation.
+
+#### Project Graduation Process: Incubating to Graduation
```
Comment (AevaOnline) [inline process/software-project-lifecycle.md:103]: as on #95, I think this section should move under ###Graduated to have a consistent layout.

File: process/software-project-lifecycle.md [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+Software focused projects that wish to move from Incubating to Graduation should open a PR confirming the following criteria:
+* Have committers from at least two organizations.
```
Comment (AevaOnline) [inline process/software-project-lifecycle.md:106]: as on #95, I propose changing this to min=3 orgs, which is the CNCF's original bar. Happy to discuss as this is a substantial change in requirements.

File: process/software-project-lifecycle.md [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+* Have achieved and maintained a [Core Infrastructure Initiative Best Practices Badge](https://bestpractices.coreinfrastructure.org/).
+* Have completed an independent and third party security audit with results published of similar scope and quality as [this example](https://github.com/envoyproxy/envoy#security-audit) which includes all critical vulnerabilities and all critical vulnerabilities need to be addressed before graduation.
```
Comment (AevaOnline) [inline process/software-project-lifecycle.md:108]: as on #95, I suggest adding words here to indicate that the TAC and/or other WGs may provide assistance with the sec audit funding, which was not true in CNCF when this was written.

File: process/working-group-lifecycle.md [hunk 5]
```diff
@@ -35,17 +37,17 @@ progress a project will move to `End-of-Life`. The details for each phase follow
...
+* Complete and request TAC approval of [README.md](https://github.com/ossf/project-template/blob/main/README.md)
 * Meet on a regular cadence. Meetings are public, recorded, and on the calendar
 * Have access to community resources (Zooms, YouTube channels, GitHub, Slack channels, etc.)
 * Can request funding/other resources (subject to TAC/GB approval)
     * NOTE: _At the time of this draft, funding and resources beyond collaboration tools have not
-    been established in the OpenSSF. Working groups should expect that their main resource is the
+    been established in the OpenSSF. WGs should expect that their main resource is the
     community contributions they are able to recruit._
 
 ## To become `Active`:
```
Comment (AevaOnline) [inline process/working-group-lifecycle.md:48]: unsurprisingly, my comments from #95 also apply here :)

**Comment by AevaOnline** [comment]
Thanks, Arnaud! I think it is helpful to see these two proposals side-by-side. Unsurprisingly, I prefer #95 -- I think the wording is clearer, more concise, aligns with the existing Charter (so avoids going back to the GB to request a change), _and_ aligns more easily with sister-organizations (thus facilitating an easier lateral move for folks coming from elsewhere, or working across multiple foundations).

I would love to hear other points of view. Do the acronyms SFP / NSFP (Non-/ Software Focused Project) make sense? 

A separate governance structure for was notably not presented in this document -- so we'll have to go write it. What would a set of governance requirements for NSFPs look like? Would it provide distinct value from the governance of SFPs and WGs? 

I think having an articulated answer to these questions would help us determine whether this 3-types-of-TIs model is necessary.

**Comment by dlorenc** [comment]
> SFP / NSFP

The acronyms make sense, but I'm not sure I understand the distinction between a NSFP and a WG. To make sure I'm on the same page - what would an example of a non software focused project be? Do we have any today?

**Comment by lehors** [comment]
> 

Yes, we have more NSFPs than SFPs according to David (and I think he's right). One such example is "The Great MFA Project" for instance, created by the Best Practices WG.

**Comment by lehors** [comment]
I've been trying to think about what the lifecycle for NSFPs could look like. Based on my knowledge of some of the existing ones I think it's got to be pretty light, closer to what we have for the WGs than what we would have for SFPs.
For instance, "The Great MFA project" was started by the Best Practices WG, executed on, and is now either completed or dormant. I don't think this requires a complicated process around it.

**Comment by dlorenc** [comment]
> For instance, "The Great MFA project" was started by the Best Practices WG, executed on, and is now either completed or dormant. I don't think this requires a complicated process around it.

Thanks, that helps.

By formalizing these, Is the idea that "The Great MFA Project" would have been done as it's own top-level NSFP rather than as just an initiative from the WG like it was? I think I'm unclear on the line between "just a thing a working group decided to do", a NSFP.

**Comment by lehors** [comment]
> > For instance, "The Great MFA project" was started by the Best Practices WG, executed on, and is now either completed or dormant. I don't think this requires a complicated process around it.
> 
> Thanks, that helps.
> 
> By formalizing these, Is the idea that "The Great MFA Project" would have been done as it's own top-level NSFP rather than as just an initiative from the WG like it was? I think I'm unclear on the line between "just a thing a working group decided to do", a NSFP.

As far as I understand there is no such line in fact. But maybe @david-a-wheeler can help us here because he's the one who's reacted most strongly against PR #95 which reserves the term "project" to SFPs on the basis that "most OpenSSF projects don't produce software today".

File: process/software-project-lifecycle.md (hunk 1 above)
Comment (inferno-chromium) [reply process/software-project-lifecycle.md:70]: Should this depend on the state of project, like early stage projects with few mvp scripts shouldnt need legal review.

File: process/software-project-lifecycle.md [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+
+#### Gets - Benefits
+*** Need Text ***
+
+#### Incubating Entry Requirements
+
+To be accepted to incubating stage, a project must meet the sandbox stage requirements plus:
+
+* Document that it is being used successfully in production by at least three independent end users which, in the TAC’s judgement, are of adequate quality and scope.
+* Have a healthy number of committers. A committer is defined as someone with the commit bit; i.e., someone who can accept contributions to some or all of the project.
```
[the remark this answers is not in the document]
Comment (inferno-chromium) [reply process/software-project-lifecycle.md:97]: +1

File: process/software-project-lifecycle.md (hunk 3 above)
Comment (inferno-chromium) [reply process/software-project-lifecycle.md:106]: Replied in #95 as well on this.

File: process/software-project-lifecycle.md (hunk 4 above)
Comment (inferno-chromium) [reply process/software-project-lifecycle.md:108]: +1

File: process/working-group-lifecycle.md (hunk 5 above)
Comment (inferno-chromium) [reply process/working-group-lifecycle.md:48]: +1

**Comment by lehors** [comment]
I've applied the same changes that I made to PR #95 in response to the comments I received.

File: process/software-project-lifecycle.md (hunk 3 above)
Comment (lehors) [reply process/software-project-lifecycle.md]: Waiting for resolution. :-)

File: process/software-project-lifecycle.md (hunk 4 above)
Comment (lehors) [reply process/software-project-lifecycle.md]: Done.

**Comment by sbtaylor15** [comment]
> 

@AevaOnline - I was thinking along the same lines.  I came up with the following:  Working Group Projects, Standalone Projects, and Standalone with SIF.   WG Projects report up to the WGs, SA Projects report up to the TAC and SA-SIF Projects report up to the TAC but have a SIF associated with them.    

I can go in any direction on this as long as it's easy to understand.

**Comment by AevaOnline** [comment]
> 
> @AevaOnline - I was thinking along the same lines around SFP / NSFP. I came up with the following: Working Group Projects, Standalone Projects, and Standalone with SIF. WG Projects report up to the WGs, SA Projects report up to the TAC and SA-SIF Projects report up to the TAC but have a SIF associated with them.
> 
> I can go in any direction on this as long as it's easy to understand.

I think we're still having some challenges teasing apart the governance relationships from the impact of a given body of work, and I believe that to be a very valuable distinction from the perspective of informing our community members and the public at-large what the different bodies within the OpenSSF do. 

I'll try to reflect my view on this in a table:

| proposed name | expected lifespan | primary output | reporting relationship | economic model |
| --- | --- | --- | --- | --- |
| **Working Group (WG)** | unbounded| not software | to the TAC | normative |
| **WG w/ SIF** | unbounded| not software | to the TAC | special fund |
| **Project** | unbounded| software | either TAC or WG | normative |
| **Project w/ SIF** | unbounded| software | to the TAC | special fund |
| **Special Interest Group (SIG)** | bounded | not software | to a WG | normative |

Examples using the above terminology:
- Vuln Disclosure WG: **WG**
- alpha/omega: **WG w/ SIF**
- great MFA distribution: **SIG**
- scorecards: **Project**
- Pyrsia: **Project** (I believe this is the team's goal -- I'm looking for confirmation that this model works for you all)

Thoughts?

File: process/software-project-lifecycle.md (hunk 1 above)
Comment (AevaOnline) [reply process/software-project-lifecycle.md]: No matter how "small" the codebase is, if there is a transfer of IP, then a legal sign-off is in everyone's best interest.

The deciding factor, in my past experience, has been whether the project is _started_ in the open (e.g., members of the foundation spin up an empty repo and start working together) or whether the project was _started_ in private and then contributed. Even "just a couple scripts" can be associated with a patent claim or a trademark. This is a legal bar, not a technical bar, and the TAC is not the deciding body on what requires legal review.

**Comment by brianf** [comment]
+1 to the clarity the table above provides.

**Comment by joshbressers** [comment]
Under this diagram could a "Project w/ SIF" fall under a "WG w/ SIF"?

I ask because I could see something like Alpha/Omega creating a lot of "Project w/ SIF" groups, that if all report to the TAC, could be a PITA.

Or would those be projects without a SIF?

**Comment by dlorenc** [comment]
This taxonomy seems correct and helpful as a set of definitions.

From a charter/governance perspective, I don't think the TAC or charter should specify or care about things that don't report directly to it, though. I'm my opinion, the TAC processes should only concern themselves with top level Working groups and Projects, allowing them to govern their own activities as they see fit.

**Comment by AevaOnline** [comment]
> Under this diagram could a "Project w/ SIF" fall under a "WG w/ SIF"?

If the nested Project is drawing its funding from the parent "WG w/ SIF"'s funding pool, then I would see it simply as a Project within that "WG w/ SIF". 

I have not seen any proposal yet to nest SIFs within SIFs.

> the TAC processes should only concern themselves with top level Working groups and Projects, allowing them to govern their own activities as they see fit.

In general, I agree, however the Project Graduation Review / Project Progression Policy needs to have a consistent bar across the whole OpenSSF to have value to the community. Because of this, I believe it needs to be a function of the TAC. That said, I see no issue with "small" projects that live within WGs.

**Comment by SecurityCRob** [comment]
+1 on the 5 group names/definitions

**Comment by AevaOnline** [comment]
In the TAC meeting today (4/19) there was general consensus from TAC members (all were present, none objected) on using the classification system represented in the above table. Steve Chin spoke in favor as well, representing the interests of the Pyrsia project.

One point was raised: the proposed use of SIG/WG is "inverted" relative to several other LF Projects, including the CNCF, Hyperledger, and others. I pointed out that aligning this would require an OpenSSF Foundation Charter amendment, which should be relatively simple to do.

I believe it is in our best interest to swap "WG" and "SIG" (making SIG the label for unbounded work and WG the label for bounded work) so as to reduce cognitive dissonance for contributors that span LF Foundation Projects. Since we are in the early stages of "rebooting" the TAC, this is the best time to make such a change.

What do others think?

**Comment by gkunz** [comment]
> I do not necessarily agree that "terminating the fund is a natural criterion for archiving"; I can envision a scenario where the primary funding for a SIF runs dry and other participants wish to continue the work. Were that to happen, I would expect the remaining participants, the TAC, and the GB would discuss how to proceed, and archiving is only one possible outcome -- but beyond that, I don't want to speculate about the process at this time.

Yes, that's true, of course. There is no need to tie the termination of the fund to the termination of the project.

**Comment by lehors** [comment]
> I believe it is in our best interest to swap "WG" and "SIG" (making SIG the label for unbounded work and WG the label for bounded work) so as to reduce cognitive dissonance for contributors that span LF Foundation Projects. Since we are in the early stages of "rebooting" the TAC, this is the best time to make such a change.
> 
> What do others think?

SGTM. I don't think there is going to be a better time to make that change.

**Comment by lehors** [comment]
I have incorporated the terminology and structure @AevaOnline proposed here into PR#95 and created PR#98 with the same structure with the names SIG and WG swapped.
I don't see any further need to keep open this PR which proposed to differentiate the various groups using the qualifiers "software focused" and "non-software focused".

**Pull request closed without merging**